[Kishendran] iP

KishendranVendarKon opened the pull request:
# Lottery-A *(List Of Tasks That Eventually Require Your Attention)*
> "Forget the lottery. Bet on yourself instead." - Brian Koslow

Do not do yourself the injustice of forgetting your tasks. Use Lottery-A and bet on yourself. It is,

* **Text-based**
* **User friendly**
* **Fast** to learn, even ***faster*** to use.

All you need to do is,

1. Download it from [here.](https://github.com/KishendranVendarKon/ip/releases/tag/A-Jar)
2. Run the jar file
3. Add your current tasks
4. And let Lottery-A manage your tasks from this day :grinning:

The best part...is that it ~actually works~ is **FREE!**

Features:

- [x] Managing Tasks (Add, delete, mark, find)
- [ ] Managing deadlines
- [ ] Reminders

If you are a Java programmer, feel free to suggest improvements. Here's the `dukeDeadline` method:

```java
    protected void dukeDeadline(String desc, Calendar cal){
        Deadline deadline = new Deadline(desc, cal);
        tasks.add(deadline);
        Ui.showAddTaskMessage(deadline.toString(), tasks.size());
    }
```

Commit:
Add Gradle support

KishendranVendarKon committed:
Greet, Echo, Exit

KishendranVendarKon committed:
Add, List

KishendranVendarKon committed:
Added Task Class

KishendranVendarKon committed:
Mark As Done

KishendranVendarKon committed:
Added Children Classes for Task Class

KishendranVendarKon committed:
ToDos, Events, Deadlines

KishendranVendarKon committed:
Automated text UI Testing

KishendranVendarKon committed:
Handle Errors

KishendranVendarKon committed:
Delete

KishendranVendarKon committed:
Transformed If Else Block to Switch Statement

KishendranVendarKon committed:
Moved core function of duke to a defined function called runDuke() called by main rather than having all the info in main

KishendranVendarKon committed:
Addition of javadoc comments

KishendranVendarKon committed:
Add save function

KishendranVendarKon committed:
Add recognition of dates as Calendar objects

KishendranVendarKon committed:
Merge branch 'branch-Level-7'

KishendranVendarKon committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java

KishendranVendarKon committed:
Use More OOP

KishendranVendarKon committed:
Organize into Packages

KishendranVendarKon committed:
Add JUnit Tests

KishendranVendarKon committed:
Update gitignore for jar file

KishendranVendarKon committed:
Add JavaDoc comments

KishendranVendarKon committed:
Follow the coding standard

KishendranVendarKon committed:
Add Find function

KishendranVendarKon committed:
Merge branch 'branch-A-JavaDoc'

KishendranVendarKon committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/task/Storage.java
#	src/main/java/duke/task/TaskList.java

KishendranVendarKon committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/command/MyParser.java

KishendranVendarKon committed:
Merge remote-tracking branch 'origin/add-gradle-support'

KishendranVendarKon committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

KishendranVendarKon committed:
Add Gradle

KishendranVendarKon committed:
Merge branch 'branch-A-Gradle'

KishendranVendarKon committed:
Use CheckStyle

KishendranVendarKon committed:
Merge branch 'branch-A-CheckStyle'

KishendranVendarKon committed:
Add GUI

KishendranVendarKon committed:
Merge branch 'branch-Level-10'

eltonyeh commented on build.gradle:
```diff
@@ -0,0 +1,62 @@
+    toolVersion = '8.29'
+}
+
+run{
```
Maybe there should be a space?
```suggestion
run {
```

hanif-kamal replied:
Perhaps there should be a space.

eltonyeh commented on src/main/java/duke/command/DialogBox.java:
```diff
@@ -0,0 +1,61 @@
+
+    public static DialogBox getDukeDialog(String text, Image img) {
+        var db = new DialogBox(text, img);
+        //db.flip();
```
Maybe this line needs to be removed?

hanif-kamal replied:
Perhaps the commented out code should be deleted.

eltonyeh commented on src/main/java/duke/command/Duke.java:
```diff
@@ -0,0 +1,118 @@
+import java.util.Calendar;
+import java.util.Scanner;
+
+import duke.task.*;
```
Should wildcard import be avoided?

eltonyeh commented on src/main/java/duke/command/Duke.java:
```diff
@@ -0,0 +1,118 @@
+ * @author Kishendran Vendar Kon (Group G05)
+ */
+public class Duke {
+    /** Contains task list. */
```
Maybe it should sound more like a noun?

eltonyeh commented on src/main/java/duke/command/Duke.java:
```diff
@@ -0,0 +1,118 @@
+
+    private MainWindow mainWindow;
+
+    /** Default Constructor. */
```
Maybe it should be three lines (instead of only one)?
```suggestion
    /** 
     *Default Constructor. 
     */
```

eltonyeh commented on src/main/java/duke/command/Duke.java:
```diff
@@ -0,0 +1,118 @@
+        mainWindow.showAddTaskMessage(event.toString(), tasks.size());
+    }
+
+    protected void dukeFind(String desc) {
```
Missing JavaDoc?

eltonyeh commented on src/main/java/duke/command/MainWindow.java:
```diff
@@ -0,0 +1,133 @@
+        try {
+            parser.parse(command, description, duke);
+        } catch (DukeException e) {
+            dialogContainer.getChildren().add(DialogBox.getDukeDialog(e.getMessage(), dukeImage));
```
Maybe it is better to separate the code into two or more lines? This question also applies to the code below.

eltonyeh commented on src/main/java/duke/command/MyParser.java:
```diff
@@ -0,0 +1,134 @@
+ *
+ * @author Kishendran Vendar Kon (Group G05)
+ */
+public class MyParser {
```
Maybe `Parser` will be a better name? It is more concise, and the `my` seems to have no special meaning.

eltonyeh commented on src/main/java/duke/command/MyParser.java:
```diff
@@ -0,0 +1,134 @@
+            duke.dukeDeadline(desc, deadlineCal);
+            break;
+        }
+        case "event": {
+            int atIndex = userDescription.indexOf("/at ");
+            if (userDescription.isBlank() || atIndex == 0) {
+                throw new DukeException("OOPS!!! The description of an event cannot be empty");
+            }
+
+            if (atIndex == -1) {
+                throw new DukeException("OOPS!!! No event time has been set. Input with '/at dd/mm/yyyy 0000'");
+            }
+
+            String desc = userDescription.substring(0, atIndex);
+            String at = userDescription.substring(atIndex + 4);
+
+            Calendar eventCal = Calendar.getInstance();
+            SimpleDateFormat eventTimeFormat = new SimpleDateFormat("d/MM/yyyy HHmm");
+
+            try {
+                eventCal.setTime(eventTimeFormat.parse(at));
+            } catch (ParseException e) {
+                throw new DukeException("OOPS!!! The date is not formatted as dd/mm/yyyy 0000");
+            }
+            duke.dukeEvent(desc, eventCal);
+            break;
+        }
+        case "find":
+            if (userDescription.isBlank()) {
+                throw new DukeException("OOPS!!! The description of find cannot be empty");
+            }
+
+            duke.dukeFind(userDescription);
+            break;
+        default:
+            throw new DukeException("OOPS!!! I'm Sorry, but I don't know what that means");
+        }
+    }
+}
+
```
Maybe this method is a bit too long? Maybe you can break down each case to separate methods?

eltonyeh commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,36 @@
+ */
+public class Deadline extends Task {
+    /** String representing dueBy date/time of Deadline. */
+    protected static SimpleDateFormat formatter = new SimpleDateFormat("MMM dd yyyy hh:mm aaa");
```
Seems that it is also `final`?

eltonyeh commented on src/main/java/duke/task/Storage.java:
```diff
@@ -0,0 +1,80 @@
+ * @author Kishendran Vendar Kon (Group G05)
+ */
+public class Storage {
+    private File f;
```
Maybe the naming should be more descriptive? 
It is an attribute that would be used in the entire class, so maybe making it more understandable is a good idea?

eltonyeh commented on src/main/java/duke/task/Storage.java:
```diff
@@ -0,0 +1,80 @@
+    public Storage(String txtFile) throws IOException {
+        String home = System.getProperty("user.home");
+        Path path = Paths.get(home, "duke", "dir");
+        txtPath = new StringBuilder().append(home).append("\\duke")
```
Doesn't `\\` only apply to Windows (correct me if I'm wrong)? I find that `File.separator` may be useful in this case.

eltonyeh commented on src/main/java/duke/task/Storage.java:
```diff
@@ -0,0 +1,80 @@
+    public ArrayList<String> load() throws IOException {
+        BufferedReader reader = new BufferedReader(new FileReader(txtPath));
+        ArrayList<String> loadedTasks = new ArrayList<>();
+        String s;
+        while ((s = reader.readLine()) != null) {
```
Maybe `s` and `reader.readline()` can be declared/called at the same place?
```suggestion
        String s = reader.readLine();
        while (s != null) {
```

eltonyeh commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,168 @@
+     */
+    public TaskList(ArrayList<String> tasks) {
+        this.toDoList = new ArrayList<>();
+        for (int i = 0; i < tasks.size(); i++) {
+            String s = tasks.get(i);
```
Maybe you can use enhanced `for` instead of the normal one?
```suggestion
        for (String s : tasks) {
```

eltonyeh commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,168 @@
+            case 'T':
+                Todo t = new Todo(s.substring(7));
+                if (s.charAt(4) == 'x') {
+                    t.markAsDone();
+                }
+                toDoList.add(t);
+                break;
+            case 'D':
+                int deadlineIndex = s.indexOf("(by:");
+                Calendar deadlineCal = Calendar.getInstance();
+
+                try {
+                    deadlineCal.setTime(formatter.parse(s.substring(deadlineIndex + 5, s.length() - 1)));
+                } catch (ParseException e) {
+                    System.out.println(e);
+                }
+
+                Deadline d = new Deadline(s.substring(7, deadlineIndex), deadlineCal);
+                if (s.charAt(4) == 'x') {
+                    d.markAsDone();
+                }
+                toDoList.add(d);
+                break;
+            case 'E':
+                int eventIndex = s.indexOf("(at:");
+
+                Calendar eventCal = Calendar.getInstance();
+
+                try {
+                    eventCal.setTime(formatter.parse(s.substring(eventIndex + 5, s.length() - 1)));
+                } catch (ParseException e) {
+                    System.out.println(e);
+                }
+
+                Event event = new Event(s.substring(7, eventIndex), eventCal);
+                if (s.charAt(4) == 'x') {
+                    event.markAsDone();
+                }
+                toDoList.add(event);
+                break;
```
Will it be better if the details in each case are written in other methods? This way, you can maintain the code more easily.

eltonyeh commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,168 @@
+                Calendar deadlineCal = Calendar.getInstance();
+
+                try {
+                    deadlineCal.setTime(formatter.parse(s.substring(deadlineIndex + 5, s.length() - 1)));
```
Maybe you can write `s.substring(deadlineIndex + 5, s.length() - 1)` as a new variable?
```suggestion
                    String s = s.substring(deadlineIndex + 5, s.length() - 1);
                    deadlineCal.setTime(formatter.parse(s));
```

eltonyeh reviewed (commented):
Most of your codes are structured logically, good job! I feel that it can be even better if you break your code down and make it easier to maintain.
BTW, I noticed that you included the `jar` file in your commits. According to the module website, maybe attaching it in a new release is better.

KishendranVendarKon committed:
Clean up code

hanif-kamal reviewed (commented):
Generally well done. Just a few minor changes needed to adhere to the coding standard.

KishendranVendarKon committed:
Use Assertions

Assertions are not used in this project.

A project not using assertions may be vulnerable to errors caused by
assumptions made by the software engineer.

Let's add assertions whenever an assumption of invariants or conditions
is made.

In this case, assertions are more appropriate then exception handling
as we are targeting mistakes made by programmer rather than the user.

KishendranVendarKon committed:
Improve Code Quality

Long methods and magic numbers are found in code.

Makes code difficult to read and follow.

Let's refactor the code to extract simpler methods
from long methods and declare constants to explain magic numbers.

Refactoring code by extracting is inline with SLAP and keeps code clean.

KishendranVendarKon committed:
Merge pull request #1 from KishendranVendarKon/branch-A-CodeQuality

Improve Code Quality

KishendranVendarKon committed:
Merge branch 'master' into branch-A-Assertions

KishendranVendarKon committed:
Merge pull request #2 from KishendranVendarKon/branch-A-Assertions

Use Assertions

KishendranVendarKon committed:
Rectify test case

KishendranVendarKon committed:
Set up CI

KishendranVendarKon committed:
Merge pull request #3 from KishendranVendarKon/branch-A-CI

Set up CI

KishendranVendarKon committed:
Update CI

KishendranVendarKon committed:
Remove ubuntu from .yml file for testing

KishendranVendarKon committed:
Revert "Remove ubuntu from .yml file for testing"

This reverts commit c3b6d5d70db891816d5e4ffcaca7c096b588b578.

KishendranVendarKon committed:
Revert "Update CI"

This reverts commit 27ebe48a79243c46bf54c29e73f103349421e443.

KishendranVendarKon committed:
Revert "Set up CI"

KishendranVendarKon committed:
Add edit command

KishendranVendarKon committed:
Merge pull request #4 from KishendranVendarKon/branch-C-Update

Add edit command

KishendranVendarKon committed:
Improve Code Quality

KishendranVendarKon committed:
Improve GUI Sizing

KishendranVendarKon committed:
Better GUI

KishendranVendarKon committed:
Further improve GUI

KishendranVendarKon committed:
Change GUI background

KishendranVendarKon committed:
Improve Code Quality

KishendranVendarKon committed:
Set theme jekyll-theme-midnight

KishendranVendarKon committed:
Set theme jekyll-theme-midnight

KishendranVendarKon committed:
Add resources

KishendranVendarKon committed:
Update Readme

KishendranVendarKon committed:
Update README.md

KishendranVendarKon committed:
Delete Ui image from docs

KishendranVendarKon committed:
Add UI Image to docs

KishendranVendarKon committed:
Finalize features

KishendranVendarKon committed:
Improve Javadoc Comments

KishendranVendarKon committed:
Resize UI
